torch/_functorch/partitioners.py

```diff
@@ -622,6 +622,74 @@ def insert_node_in_graph(node):
     return new_gm
 
 
+def apply_graphsafe_rng_functionalization(
+    fw_module: torch.fx.GraphModule,
+    bw_module: torch.fx.GraphModule,
+    fw_node: torch.fx.Node,
+    bw_node: torch.fx.Node,
+    device: torch.device,
+    rng_count: int,
+    last_fwd_input: torch.fx.Node,
+    last_bwd_input: torch.fx.Node,
+):
+    """Applies graph-safe RNG functionalization to forward and backward nodes in a CUDA device context.
+
+    This function modifies the forward and backward computation graphs by:
+    1. Creating RNG state placeholders for both passes
+    2. Updating the forward node to use graph-safe RNG state
+    3. Updating the backward node to use graph-safe RNG state
+    """
+    device_idx = device.index
+    assert device_idx is not None
+    fw_graph = fw_module.graph
+    bw_graph = bw_module.graph
+    graphsafe_run_with_rng_state = torch._prims.rng_prims.graphsafe_run_with_rng_state
+
+    # Handle forward pass
+    with fw_module.graph.inserting_after(last_fwd_input):
+        fwd_rng_state = fw_module.graph.placeholder(f"fwd_rng_state_{rng_count}")
```

Comment:
I guess adding one graph input per rng is pretty inevitable. I wonder if we're going to see any compile time perf problems down the line, though, for models that have larger numbers of rng ops in them

Reply:
Yeaaaa. maybe we can check to see if cudagraphs is enabled. although its a little weird for that to change numerics. but maybe thats inevitable.

Reply:
Other question for reviewers:

> This does change numerics, bc the rng of the op is now taken from the input rng state instead of whatever the rng would be midway through running the graph. Technically, we only need this for cuda graph. But, I'd prefer to not have a rng divergence just for cudagraph. I am making it respect fallback_random.

Not sure if you or @anijain2305 have any thoughts here.

Reply:
> (2) my understanding is that... this is not fundamental? As in, we could theoretically change the inductor rng logic to mirror eager numerics, but this is a large chunk of work. Is that right?

I honestly don't know if this is exactly possible, at the same perf. I know that today triton's rng throws away 3/4 rng values or something. it also obviously has a different programming model. maybe you can, but i can't confidently say today.

3) - yea this is correct. and yea, that's reasonable, but i'm not sure we'll ever have to deliberate on that honestly. but we can discuss if we get there